### PR TITLE
[collectd 6] Write OpenTelemetry: New plugin to export metrics via OTLP over gRPC.

### PR DESCRIPTION
**Pull request by octo**
This adds a new plugin that exports metrics to an [OpenTelemetry Collector](https://opentelemetry.io/docs/collector/) using the *OpenTelemetry Protocol* (OTLP) over gRPC.

This is intended as an MVP and comes with many limitations, e.g.:

* Communication is synchronous and blocks the write thread
* TLS is not yet supported
* ~The [Resource field](https://opentelemetry.io/docs/concepts/resources/) is unset (→ see #4187)~

ChangeLog: Write OpenTelemetry: A new plugin for exporting metrics using the OpenTelemetry Protocol over gRPC has been added.

### CODE REVIEW

File: src/write_open_telemetry.cc
Comment (eero-t) [inline src/write_open_telemetry.cc:389]: Btw. I'm not that familiar how plugin work.  Why these could not be registered in `module_register()`?

**Review by eero-t — COMMENTED** [review]
Somebody more familiar e.g. with collectd AVL API, needs to check the actual plugin code.  Other changes look OK.

**Comment by eero-t** [comment]
Btw. using `[collectd 6]` subject prefix for PRs going into `collectd-6.0` branch, like the earlier PRs for it do, would be nice.

(Makes it e.g. easier to see at glance where merged PRs went, without needing to use `base:collectd-6.0` filter.)

File: src/write_open_telemetry.cc
Comment (octo) [reply src/write_open_telemetry.cc:389]: This is the common pattern used when – depending on the configuration – multiple callbacks may be registered, e.g. to send metrics to multiple collectors. The information which callbacks to register is only available during the configuration phase, and therefore we can't register the callbacks in `module_register()`.

**Comment by octo** [comment]
Rebased and updated to take advantage of #4187.

**Comment by eero-t** [comment]
Somebody more familiar e.g. with collectd AVL API, needs to check the actual `write_open_telemetry.cc` plugin code.  

Updates on the other code (still) look OK.

**Comment by octo** [comment]
I have created a new utility, `resource_metrics`, which is also used by the OTLP JSON code (#4188). I'm happy to create a separate PR just for that utility to make the review simpler. Just let me know.

With this new utility, the AVL trees are no longer used. Does that help with the review?

**Comment by eero-t** [comment]
Please split workflow update and bionic dropping to another PR.

**Comment by octo** [comment]
> Please split workflow update and bionic dropping to another PR.

I moved the dropping of Bionic into #4201.

I think the other workflow update (fetching submodules) is directly related to this change – it adds the `opentelemetry-proto` submodule – and should remain.

**Comment by eero-t** [comment]
`clang-format` and `release_ready` CI checks fail:
```
...
  PROTOC   collectd.pb.cc
make[1]: *** No rule to make target 'opentelemetry-proto/opentelemetry/proto/collector/metrics/v1/metrics_service.proto', needed by 'opentelemetry/proto/collector/metrics/v1/metrics_service.grpc.pb.cc'.  Stop.
make[1]: *** Waiting for unfinished jobs....
  PROTOC   types.pb.cc
/tmp/cirrus-ci-build/src/liboconfig/parser.y: warning: fix-its can be applied.  Rerun with option '--update'. [-Wother]
updating src/liboconfig/parser.h
```

**Comment by octo** [comment]
> `clang-format` and `release_ready` CI checks fail:

* [x] clang-format fixed
* [x] parser.y updates are in #4202

**Comment by eero-t** [comment]
>     * [x]  parser.y updates are in [[collectd 6] liboconfig: Apply automatic fixes. #4202](https://github.com/collectd/collectd/pull/4202)

Fail was due to `No rule to make target ...` error (`parser.y` thing was just warning).

**Comment by octo** [comment]
> Fail was due to `No rule to make target ...` error

CirrusCI was missing the Git submodule. Fixed.

**Comment by octo** [comment]
Hi @eero-t, I'd like to merge this soon to unblock a collectd 6 MVP. Are there any changes you'd like to see before that?

File: src/daemon/metric.c
Comment (eero-t) [inline src/daemon/metric.c:571]: `metric_family_clone_shallow()` also do this check, so it could be dropped from here:
```suggestion
  metric_family_t *ret = metric_family_clone_shallow(fam);
```
But compiler should optimize it away, so you can just resolve this, if you'd rather keep it.

File: src/utils/format_open_telemetry/format_open_telemetry.cc
Comment (eero-t) [inline src/utils/format_open_telemetry/format_open_telemetry.cc:76]: Maybe `FATAL()`?

File: src/utils/format_open_telemetry/format_open_telemetry.cc
Comment (eero-t) [inline src/utils/format_open_telemetry/format_open_telemetry.cc:87]:
```suggestion
    metric_to_number_data_point(dp, m);
```

File: src/utils/format_open_telemetry/format_open_telemetry.cc
Comment (eero-t) [inline src/utils/format_open_telemetry/format_open_telemetry.cc:122]: `FATAL()`?

File: src/write_open_telemetry.cc
Comment (eero-t) [inline src/write_open_telemetry.cc:134]: I think this would be easier to read with args reversed:
```suggestion
    if (now < (cb->staged_time + timeout))
      return 0;
```

File: src/daemon/metric.c
Comment (octo) [reply src/daemon/metric.c]: A good rule of thumb is that less code contains fewer bugs. Committed.

File: src/utils/format_open_telemetry/format_open_telemetry.cc
Comment (octo) [reply src/utils/format_open_telemetry/format_open_telemetry.cc]: Good point, I unified this now.

File: src/utils/format_open_telemetry/format_open_telemetry.cc
Comment (octo) [reply src/utils/format_open_telemetry/format_open_telemetry.cc:76]: Ugh, yeah, that's still in there …

I think `METRIC_TYPE_UNTYPED` as is exists today has problems:

* It is not zero. So it cannot used to detect broken plugins that fail to initialize the structure properly.
* It makes it appear as if untyped metrics would somehow be legal. Maybe the renaming it to "uninitialized" would be better.

That is for another PR though. For now, I'll make sure to skip metric families entirely.

Comment (octo) [reply src/utils/format_open_telemetry/format_open_telemetry.cc]: Done

File: src/utils/format_open_telemetry/format_open_telemetry.cc
Comment (octo) [reply src/utils/format_open_telemetry/format_open_telemetry.cc]: As above, I have implemented graceful handling of this situation. The case must remain, though, otherwise compilers complain about an unhandled value of the enum.

**Comment by octo** [comment]
Thanks @eero-t, I think I have addressed all your comments. Thank you so much for the review!

File: src/write_open_telemetry.cc
Comment (eero-t) [inline src/write_open_telemetry.cc:157]: MetricsService stub is just reset, it and channel gets never freed?  Would that need some suppression for Valgrind memcheck leak checker?

**Review by eero-t — APPROVED** [review]
On cursory review this looks now OK => approved.

I would suggest running it under Valgrind (memcheck & helgrind tools) with multiple plugins providing metrics, before merging though.

File: src/write_open_telemetry.cc
Comment (eero-t) [reply src/write_open_telemetry.cc:157]: If you enable `--leak-check=full`, does that "definitely lost" originate from `collectd`?

Comment (octo) [reply src/write_open_telemetry.cc:157]: I fixed a minor leak (e93000d724f43c6718c3665c4254dd7d39b66e0c, the callback's name wasn't freed). The remainder are outside of our control:

```
==172573== 40 bytes in 1 blocks are definitely lost in loss record 108 of 227
==172573==    at 0x48407B4: malloc (vg_replace_malloc.c:381)
==172573==    by 0x4002447: malloc (rtld-malloc.h:56)
==172573==    by 0x4002447: _dl_map_object_deps (dl-deps.c:422)
==172573==    by 0x400B8EC: dl_open_worker_begin (dl-open.c:592)
==172573==    by 0x49DEE09: _dl_catch_exception (dl-error-skeleton.c:208)
==172573==    by 0x400B045: dl_open_worker (dl-open.c:782)
==172573==    by 0x49DEE09: _dl_catch_exception (dl-error-skeleton.c:208)
==172573==    by 0x400B437: _dl_open (dl-open.c:884)
==172573==    by 0x4915347: dlopen_doit (dlopen.c:56)
==172573==    by 0x49DEE09: _dl_catch_exception (dl-error-skeleton.c:208)
==172573==    by 0x49DEEBE: _dl_catch_error (dl-error-skeleton.c:227)
==172573==    by 0x4914E36: _dlerror_run (dlerror.c:138)
==172573==    by 0x49153F8: dlopen_implementation (dlopen.c:71)
==172573==    by 0x49153F8: dlopen@@GLIBC_2.34 (dlopen.c:81)
==172573==
==172573== 2,816 bytes in 8 blocks are possibly lost in loss record 212 of 227
==172573==    at 0x484220F: operator new[](unsigned long) (vg_replace_malloc.c:640)
==172573==    by 0x63BE3FF: ??? (in /usr/lib/x86_64-linux-gnu/libabsl_random_internal_pool_urbg.so.20220623.0.0)
==172573==    by 0x5CEE4CE: void absl::debian3::base_internal::CallOnceImpl<void (&)()>(std::atomic<unsigned int>*, absl::debian3::base_internal::SchedulingMode, void (&)()) (in /usr/lib/x86_64-linux-gnu/libabsl_malloc_internal.so.20220623.0.0)
==172573==    by 0x63BE5F7: ??? (in /usr/lib/x86_64-linux-gnu/libabsl_random_internal_pool_urbg.so.20220623.0.0)
==172573==    by 0x63BE7E0: absl::debian3::random_internal::RandenPool<unsigned char>::Fill(absl::debian3::Span<unsigned char>) (in /usr/lib/x86_64-linux-gnu/libabsl_random_internal_pool_urbg.so.20220623.0.0)
==172573==    by 0x5902613: grpc_core::BackOff::BackOff(grpc_core::BackOff::Options const&) (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172573==    by 0x57E2000: grpc_core::PollingResolver::PollingResolver(grpc_core::ResolverArgs, grpc_core::ChannelArgs const&, grpc_core::Duration, grpc_core::BackOff::Options, grpc_core::TraceFlag*) (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172573==    by 0x57C560C: ??? (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172573==    by 0x57C5CD1: ??? (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172573==    by 0x59A63DA: grpc_core::ResolverRegistry::CreateResolver(absl::debian3::string_view, grpc_core::ChannelArgs const&, grpc_pollset_set*, std::shared_ptr<grpc_core::WorkSerializer>, std::unique_ptr<grpc_core::Resolver::ResultHandler, std::default_delete<grpc_core::Resolver::ResultHandler> >) const (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172573==    by 0x572099E: grpc_core::ClientChannel::CreateResolverLocked() (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172573==    by 0x5720CF1: grpc_core::ClientChannel::TryToConnectLocked() (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
```

Comment (eero-t) [reply src/write_open_telemetry.cc:157]: You could try e.g. `--num-callers=30` to see whether "possibly lost" in _grpc_ leads to something that could be  due to something `collectd` does.

Comment (octo) [reply src/write_open_telemetry.cc:157]: Here you go:

```
==172707== 2,816 bytes in 8 blocks are possibly lost in loss record 212 of 227
==172707==    at 0x484220F: operator new[](unsigned long) (vg_replace_malloc.c:640)
==172707==    by 0x63BE3FF: ??? (in /usr/lib/x86_64-linux-gnu/libabsl_random_internal_pool_urbg.so.20220623.0.0)
==172707==    by 0x5CEE4CE: void absl::debian3::base_internal::CallOnceImpl<void (&)()>(std::atomic<unsigned int>*, absl::debian3::base_internal::SchedulingMode, void (&)()) (in /usr/lib/x86_64-linux-gnu/libabsl_malloc_internal.so.20220623.0.0)
==172707==    by 0x63BE5F7: ??? (in /usr/lib/x86_64-linux-gnu/libabsl_random_internal_pool_urbg.so.20220623.0.0)
==172707==    by 0x63BE7E0: absl::debian3::random_internal::RandenPool<unsigned char>::Fill(absl::debian3::Span<unsigned char>) (in /usr/lib/x86_64-linux-gnu/libabsl_random_internal_pool_urbg.so.20220623.0.0)
==172707==    by 0x5902613: grpc_core::BackOff::BackOff(grpc_core::BackOff::Options const&) (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x57E2000: grpc_core::PollingResolver::PollingResolver(grpc_core::ResolverArgs, grpc_core::ChannelArgs const&, grpc_core::Duration, grpc_core::BackOff::Options, grpc_core::TraceFlag*) (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x57C560C: ??? (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x57C5CD1: ??? (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x59A63DA: grpc_core::ResolverRegistry::CreateResolver(absl::debian3::string_view, grpc_core::ChannelArgs const&, grpc_pollset_set*, std::shared_ptr<grpc_core::WorkSerializer>, std::unique_ptr<grpc_core::Resolver::ResultHandler, std::default_delete<grpc_core::Resolver::ResultHandler> >) const (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x572099E: grpc_core::ClientChannel::CreateResolverLocked() (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x5720CF1: grpc_core::ClientChannel::TryToConnectLocked() (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x5966FCF: grpc_core::WorkSerializer::WorkSerializerImpl::DrainQueueOwned() (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x596755B: grpc_core::WorkSerializer::Run(std::function<void ()>, grpc_core::DebugLocation const&) (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x5716A0F: ??? (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x4F6AABC: closure_impl::closure_wrapper(void*, absl::debian3::Status) (in /usr/lib/x86_64-linux-gnu/libgrpc++.so.1.51.1)
==172707==    by 0x597C685: grpc_core::ExecCtx::Flush() (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x4F55610: grpc_core::ExecCtx::~ExecCtx() (in /usr/lib/x86_64-linux-gnu/libgrpc++.so.1.51.1)
==172707==    by 0x5A427F2: grpc_call_start_batch (in /usr/lib/x86_64-linux-gnu/libgrpc.so.29.0.0)
==172707==    by 0x4EABC46: grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal::CallOpSendMessage, grpc::internal::CallOpRecvInitialMetadata, grpc::internal::CallOpRecvMessage<google::protobuf::MessageLite>, grpc::internal::CallOpClientSendClose, grpc::internal::CallOpClientRecvStatus>::ContinueFillOpsAfterInterception() (call_op_set.h:971)
==172707==    by 0x4EB0646: PerformOps (call.h:67)
==172707==    by 0x4EB0646: grpc::internal::BlockingUnaryCallImpl<google::protobuf::MessageLite, google::protobuf::MessageLite>::BlockingUnaryCallImpl(grpc::ChannelInterface*, grpc::internal::RpcMethod const&, grpc::ClientContext*, google::protobuf::MessageLite const&, google::protobuf::MessageLite*) (client_unary_call.h:80)
==172707==    by 0x4EA55AF: BlockingUnaryCall<opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceRequest, opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceResponse, google::protobuf::MessageLite, google::protobuf::MessageLite> (client_unary_call.h:50)
==172707==    by 0x4EA55AF: opentelemetry::proto::collector::metrics::v1::MetricsService::Stub::Export(grpc::ClientContext*, opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceRequest const&, opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceResponse*) (metrics_service.grpc.pb.cc:43)
==172707==    by 0x4EA3646: ??? (write_open_telemetry.cc:103)
==172707==    by 0x4EA423E: ot_flush_nolock (write_open_telemetry.cc:137)
==172707==    by 0x4EA423E: ot_flush_nolock (write_open_telemetry.cc:124)
==172707==    by 0x4EA423E: ??? (write_open_telemetry.cc:178)
==172707==    by 0x119CB2: plugin_flush (plugin.c:1877)
==172707==    by 0x118702: plugin_flush_timeout_callback (plugin.c:1264)
==172707==    by 0x116AD0: plugin_read_thread (plugin.c:529)
==172707==    by 0x4919043: start_thread (pthread_create.c:442)
==172707==    by 0x499887F: clone (clone.S:100)
```

`absl::debian3::base_internal::CallOnceImpl` sounds to me like this is initializing some memory at start-up (a random pool?) and never cleans it up. Letting the binary run for longer yields the same 2816 bytes.

Comment (eero-t) [reply src/write_open_telemetry.cc:157]: I agree, this can be ignored.

(It's good to know that it was `MetricsService::Stub::Export()` call in collectd that triggered though.)

Comment (octo) [reply src/write_open_telemetry.cc:157]: Yeah, probably lazy initialization …

**Pull request merged**
